apps/backend/src/libs/hooks/check-user-permissions.hook.ts

```diff
@@ -1,13 +1,22 @@
-import { type FastifyRequest } from "fastify";
-
-import { ExceptionMessage } from "~/libs/enums/enums.js";
+import { ExceptionMessage, type PermissionKey } from "~/libs/enums/enums.js";
 import { checkHasPermission } from "~/libs/helpers/helpers.js";
-import { type APIPreHandler } from "~/libs/modules/controller/controller.js";
+import {
+	type APIHandlerOptions,
+	type APIPreHandler,
+} from "~/libs/modules/controller/controller.js";
 import { HTTPCode, HTTPError } from "~/libs/modules/http/http.js";
+import { type UserAuthResponseDto } from "~/modules/users/users.js";
+
+import { type ValueOf } from "../types/types.js";
 
-const checkUserPermissions = (routePermissions: string[]): APIPreHandler => {
-	return (request: FastifyRequest, _, done): void => {
-		const { user } = request;
+const checkUserPermissions = (
+	permissions: ValueOf<typeof PermissionKey>[],
+	getProjectId?: (options: APIHandlerOptions) => number | undefined,
+): APIPreHandler => {
+	return (options, done): void => {
+		const user = options.user as null | UserAuthResponseDto;
```

Comment:
we will only call this on endpoints which require user to be present
```suggestion
		const user = options.user as UserAuthResponseDto;
```